### ISSUE TITLE
Simulation explodes after adding a TorchForce that is identical to zero

### ISSUE THREAD

**Issue by xqding** [openmm/openmm-torch#42]
The following script simulates a simple system with 10 particles. The system has two forces: a simple harmonic bond force and a TorchForce. The forces applied on particles via the TorchForce is made into zeros, so the system with the TorchForce should be identical to that without the TorchForce. Without the TorchForce, simulations of the system is stable. After adding the TorchForce, simulation of the system explodes. This is strange because the TorchForce contributes nothing to the system.

```python
import torch
torch.set_default_dtype(torch.double)
import numpy as np
import simtk.openmm.app  as app
import simtk.openmm as omm
import simtk.unit as unit
from openmmtorch import TorchForce
import copy

class NNForceDihedral(torch.nn.Module):
    def __init__(self, num_dihedrals):
        super(NNForceDihedral, self).__init__()

	self.num_dihedrals = num_dihedrals
	self.fc1 = torch.nn.Linear(2*num_dihedrals, num_dihedrals)
        self.fc2 = torch.nn.Linear(num_dihedrals, 1, bias = False)

    def forward(self, dihedrals):
        x = torch.cat((torch.cos(dihedrals), torch.sin(dihedrals)), -1)
	x = torch.tanh(self.fc1(x))
        output = self.fc2(x)
        output = torch.squeeze(output)
	return output

class ForceModule(torch.nn.Module):
    def __init__(self, num_particles):
        super(ForceModule, self).__init__()
	self.num_particles = num_particles
	self.nnforce_dihedral = NNForceDihedral(num_particles - 3)

	self.index = torch.nn.Parameter(
            torch.LongTensor( [[i, i+1, i+2, i+3] for i in range(num_particles - 3)]),
            requires_grad = False
	)

    def forward(self, positions):
        xyz = torch.unsqueeze(positions, 0)

        xyz_i = torch.index_select(xyz, 1, self.index[:, 0])
        xyz_j = torch.index_select(xyz, 1, self.index[:, 1])
        xyz_k = torch.index_select(xyz, 1, self.index[:, 2])
        xyz_l = torch.index_select(xyz, 1, self.index[:, 3])

        b1 = xyz_i - xyz_j
        b2 = xyz_j - xyz_k
        b3 = xyz_l - xyz_k

        b1_cross_b2 = torch.cross(b1, b2)
        b3_cross_b2 = torch.cross(b3, b2)

        cos_d = torch.norm(b2, dim = -1)*torch.sum(b1_cross_b2*b3_cross_b2, -1)
        sin_d = torch.sum(b2*torch.cross(b3_cross_b2, b1_cross_b2), -1)

        dihedrals = torch.atan2(sin_d, cos_d)
        potential_energy = self.nnforce_dihedral(dihedrals)*torch.tensor(0.0)

        return potential_energy

xyz = torch.tensor([[ 1.357623  , -1.3806132 , -1.2257496 ],
                    [ 1.0385588 , -1.3198054 , -1.4279023 ],
                    [ 0.7484111 , -1.471794  , -1.6183405 ],
                    [ 0.6739166 , -1.2910188 , -1.9467432 ],
                    [ 0.33078226, -1.451046  , -1.9793787 ],
                    [ 0.15436828, -1.2586378 , -1.68369   ],
                    [ 0.42108235, -1.0122279 , -1.616695  ],
                    [ 0.4789558 , -1.1335697 , -1.2618917 ],
                    [ 0.8091399 , -1.080018  , -1.087149  ],
                    [ 0.97696364, -1.3592714 , -0.9014565 ]], requires_grad = True)

system_01 = omm.System()
system_0 = omm.System()
system_1 = omm.System()
for _ in range(10):
    system_01.addParticle(100)
    system_0.addParticle(100)
    system_1.addParticle(100)


harmonic_bond_force = omm.HarmonicBondForce()
for i in range(9):
    harmonic_bond_force.addBond(i, i+1, 0.38, 7e4)

harmonic_bond_force.setForceGroup(0)
system_01.addForce(copy.copy(harmonic_bond_force))
system_0.addForce(copy.copy(harmonic_bond_force))

force_module = ForceModule(10)

scripted_force_module = torch.jit.script(force_module)
scripted_force_module.save(f"./output/test/force_module.pth")
nnforce = TorchForce(f"./output/test/force_module.pth")
nnforce.setForceGroup(1)
system_01.addForce(copy.copy(nnforce))
system_1.addForce(copy.copy(nnforce))


T = 300*unit.kelvin
fricCoef = 1./unit.picoseconds
stepsize = 2. * unit.femtoseconds

## simulation with both forces
print("="*80)
print("Simulate system with both forces")
print("="*80)
integrator_01 = omm.LangevinMiddleIntegrator(T, fricCoef, stepsize)
platform_01 = omm.Platform.getPlatformByName('Reference')
context_01 = omm.Context(system_01, integrator_01, platform_01)
context_01.setPositions(xyz.data.numpy())

for i in range(1000):
    integrator_01.step(1)

    state = context_01.getState(getEnergy = True, getForces = True, groups = set([0]))
    bond_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    bond_force = state.getForces()
    bond_force = np.array(bond_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    state = context_01.getState(getEnergy = True, getForces = True, groups = set([1]))
    torch_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    torch_force = state.getForces()
    torch_force = np.array(torch_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    if (i + 1) % 50 == 0:
        print(f"step: {i:>5}, bond_energy: {bond_energy:>10.2f} kJ/mol, torch_energy: {torch_energy:>10.2f} kJ/mol, bond_force[0,0]:", bond_force[0,0], 'torch\
_force[0,0]: ', torch_force[0,0])

## simulation with only harmonic force
print("="*80)
print("Simulate system with just the harminic bond force")
print("="*80)

integrator_0 = omm.LangevinMiddleIntegrator(T, fricCoef, stepsize)
platform_0 = omm.Platform.getPlatformByName('Reference')
context_0 = omm.Context(system_0, integrator_0, platform_0)
context_0.setPositions(xyz.data.numpy())

for i in range(1000):
    integrator_0.step(1)

    state = context_0.getState(getEnergy = True, getForces = True, groups = set([0]))
    bond_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    bond_force = state.getForces()
    bond_force = np.array(bond_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    if (i + 1) % 50 == 0:
        print(f"step: {i:>5}, bond_energy: {bond_energy:>10.2f} kJ/mol, bond_force[0,0]:", bond_force[0,0])

## simulation with only Torch force
print("="*80)
print("Simulate system with just the Torch force")
print("="*80)

integrator_1 = omm.LangevinMiddleIntegrator(T, fricCoef, stepsize)
platform_1 = omm.Platform.getPlatformByName('Reference')
context_1 = omm.Context(system_1, integrator_1, platform_1)
context_1.setPositions(xyz.data.numpy())

for i in range(1000):
    integrator_1.step(1)

    state = context_1.getState(getEnergy = True, getForces = True, groups = set([1]))
    torch_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    torch_force = state.getForces()
    torch_force = np.array(torch_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    if (i + 1) % 50 == 0:
        print(f"step: {i:>5}, torch_energy: {torch_energy:>10.2f} kJ/mol, torch_force[0,0]: ", torch_force[0,0])
```

**Comment by peastman** [comment]
Is it possible the potential energy is nan?  0\*nan = nan, so multiplying by zero doesn't protect against that.

Try changing `forward()` to just return `torch.tensor(0.0)`.  Does that fix the error?

When you say the system explodes, what exactly does that mean?

**Comment by xqding** [comment]
> Is it possible the potential energy is nan? 0*nan = nan, so multiplying by zero doesn't protect against that.

No. The potential energy is not nan.

> Try changing forward() to just return torch.tensor(0.0). Does that fix the error?

It will make the output of forward not depend on positions and `backward()` fails when running `integrator.step()`

> When you say the system explodes, what exactly does that mean?

There are energy and forces when I run the simulation with both forces.
```
step:    49, bond_energy:       9.78 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: -172.5440207426281 torch_force[0,0]:  -0.0
step:    99, bond_energy:      54.97 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: -420.12118786591594 torch_force[0,0]:  -0.0
step:   149, bond_energy:     163.69 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: -951.7646243542338 torch_force[0,0]:  -0.0
step:   199, bond_energy:     218.56 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: -655.8039170213162 torch_force[0,0]:  -0.0
step:   249, bond_energy:     331.12 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: -437.7507816703522 torch_force[0,0]:  -0.0
step:   299, bond_energy:     769.24 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 309.1108546809626 torch_force[0,0]:  -0.0
step:   349, bond_energy:    1623.86 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 1433.0469911168857 torch_force[0,0]:  -0.0
step:   399, bond_energy:    3072.85 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 3615.10910620857 torch_force[0,0]:  -0.0
step:   449, bond_energy:    5227.31 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 6557.018456337815 torch_force[0,0]:  -0.0
step:   499, bond_energy:    8070.16 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 9386.699166534883 torch_force[0,0]:  -0.0
step:   549, bond_energy:   10989.10 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 12467.44979006325 torch_force[0,0]:  -0.0
step:   599, bond_energy:   13379.08 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 14124.743691495256 torch_force[0,0]:  -0.0
step:   649, bond_energy:   15723.78 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 15339.162440984446 torch_force[0,0]:  -0.0
step:   699, bond_energy:   17870.30 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 15321.909408580654 torch_force[0,0]:  -0.0
step:   749, bond_energy:   20172.02 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 14036.240032495009 torch_force[0,0]:  -0.0
step:   799, bond_energy:   22963.44 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 11968.834318948691 torch_force[0,0]:  -0.0
step:   849, bond_energy:   24892.63 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 9072.701338418708 torch_force[0,0]:  -0.0
step:   899, bond_energy:   26804.69 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 6840.900627499033 torch_force[0,0]:  -0.0
step:   949, bond_energy:   28595.14 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 6137.2612080825165 torch_force[0,0]:  -0.0
step:   999, bond_energy:   30047.57 kJ/mol, torch_energy:       0.00 kJ/mol, bond_force[0,0]: 6321.2805898622555 torch_force[0,0]:  -0.0
```
In contrast, if I remove the Torchforce from the system, the output is:
```
step:    49, bond_energy:       8.67 kJ/mol, bond_force[0,0]: 212.660281480979
step:    99, bond_energy:       4.71 kJ/mol, bond_force[0,0]: -267.9180875887757
step:   149, bond_energy:      10.06 kJ/mol, bond_force[0,0]: 136.14503346733807
step:   199, bond_energy:      15.78 kJ/mol, bond_force[0,0]: -143.77560006592444
step:   249, bond_energy:       6.82 kJ/mol, bond_force[0,0]: -41.02085184376862
step:   299, bond_energy:       3.31 kJ/mol, bond_force[0,0]: -90.79875374224873
step:   349, bond_energy:      11.95 kJ/mol, bond_force[0,0]: 23.924449219012708
step:   399, bond_energy:      13.65 kJ/mol, bond_force[0,0]: 468.31080993742523
step:   449, bond_energy:       9.45 kJ/mol, bond_force[0,0]: -440.58246379917875
step:   499, bond_energy:       9.45 kJ/mol, bond_force[0,0]: 292.1540536486273
step:   549, bond_energy:       8.33 kJ/mol, bond_force[0,0]: -84.95773688188685
step:   599, bond_energy:       9.44 kJ/mol, bond_force[0,0]: 36.163446661253175
step:   649, bond_energy:      11.89 kJ/mol, bond_force[0,0]: 13.199135002729014
step:   699, bond_energy:      11.03 kJ/mol, bond_force[0,0]: -118.21520585614694
step:   749, bond_energy:       4.02 kJ/mol, bond_force[0,0]: 68.1349399329583
step:   799, bond_energy:       4.93 kJ/mol, bond_force[0,0]: -13.9101667260323
step:   849, bond_energy:      10.25 kJ/mol, bond_force[0,0]: -14.770865965299054
step:   899, bond_energy:       7.84 kJ/mol, bond_force[0,0]: 3.9590041471343818
step:   949, bond_energy:      10.84 kJ/mol, bond_force[0,0]: 10.869804790126024
step:   999, bond_energy:      12.38 kJ/mol, bond_force[0,0]: -1.8712910784208492
```

If I remove the harmonic force, the output is
```
step:    49, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:    99, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   149, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   199, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   249, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   299, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   349, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   399, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   449, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   499, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   549, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   599, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   649, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   699, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   749, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   799, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   849, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   899, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   949, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
step:   999, torch_energy:       0.00 kJ/mol, torch_force[0,0]:  0.0
```

With both forces, the energy of the simulation would increase rapidly. With only either one of the two forces, the simulation looks fine.

**Comment by raimis** [comment]
@xqding try to change the order you add the forces to the system, i.e. `TorchForce` first, `HarmonicBondForce` second.

**Comment by xqding** [comment]
> @xqding try to change the order you add the forces to the system, i.e. `TorchForce` first, `HarmonicBondForce` second.

Thanks for the suggestion. It does fix the problem. Why would the order of adding forces to the system make a difference here?

In case it is useful for debugging, I updated the script in my original comment to reproduce the issue.

**Comment by raimis** [comment]
> Why would the order of adding forces to the system make a difference here?
I guess that this is because OpenMM and PyTorch create separate CUDA contexts (ideally they should use the same). We tried to mitigate this incompatibility (#16, #17) , but probably it isn't enough. Changing the order of the forces, changes the order the contexts are initialized.

**Comment by xqding** [comment]
@raimis Thanks for the explanation!

**Comment by dominicrufa** [comment]
i think i encountered this, too. is there a simple way to change the order of `Force` objects in a `System` without creating a new `System` altogether? an alternative is to place the `TorchForce` into a different `ForceGroup` than the other `Force` objects.

**Comment by raimis** [comment]
I think that I have tracked down the cause of this issues (https://github.com/openmm/openmm/pull/3256).

**Comment by peastman** [comment]
In https://github.com/openmm/openmm/pull/3258 I reworked context management as you suggested.  I still can't get this to work though.  I removed the calls to `setAsCurrent()` at https://github.com/openmm/openmm-torch/blob/76f55e7f43785d2331bb9366ffcd38a497609662/platforms/cuda/src/CudaTorchKernels.cpp#L70 and https://github.com/openmm/openmm-torch/blob/76f55e7f43785d2331bb9366ffcd38a497609662/platforms/cuda/src/CudaTorchKernels.cpp#L108 and instead push OpenMM's CUDA context just where it's directly using CUDA.  I still get the same error though.  Do you have any idea what I'm missing?

**Comment by raimis** [comment]
The fix (#46) is merged. @xqding could you confirm it works for you?

**Comment by raimis** [comment]
Adapting https://github.com/openmm/openmm/pull/3258 may take some time. @peastman we should make a release so at least the reference platform works correctly.

**Comment by xqding** [comment]
> The fix (#46) is merged. @xqding could you confirm it works for you?

@raimis Sure. Thanks for the fix. I assume I will need to compile it from source in order to incorporate the fix. It might take me some time because I have had some trouble with compiling from source on the cluster I am using. I will try to test it over the weekend and let you know.

**Comment by peastman** [comment]
I just created the release.  Conda-forge should automatically generate a PR to build new packages within a few hours.

**Comment by peastman** [comment]
In the above code, I changed Reference to CUDA.  That caused it to fail with the exception

```
      def addmm(self: Tensor, mat1: Tensor, mat2: Tensor, beta: number = 1.0, alpha: number = 1.0):
          return self + mat1.mm(mat2)
                        ~~~~~~~ <--- HERE

      def batch_norm(input : Tensor, running_mean : Optional[Tensor], running_var : Optional[Tensor], training : bool, momentum : float, eps : float) -> Tensor:
RuntimeError: expected scalar type Float but found Double
```

The fix is to add `.to(torch.float32)` to the two Linear layers.  Once I do that, the script runs without problem even when not using the changes from https://github.com/openmm/openmm/pull/3258.  Do we have an example of something that fails without the CUDA changes?

**Comment by raimis** [comment]
Yes, this doesn't capture the CUDA problems and I don't have a small example to demonstrate them.

**Comment by raimis** [comment]
Also, still there are no packages for the version 0.4.

**Comment by peastman** [comment]
For some reason it didn't create the PR.  I just created one by hand.  https://github.com/conda-forge/openmm-torch-feedstock/pull/10

**Comment by peastman** [comment]
The packages should be up now.

**Comment by raimis** [comment]
OpenMM-Torch 0.4 is out!

@xqding can you verify the fix.

@peastman regarding https://github.com/openmm/openmm/pull/3258, I'll open a separate issue, when I'll manage to make a simple and deterministic example.

**Comment by xqding** [comment]
Thanks for creating the release, which makes installing much easier. I ran the test using OpenMM-Torch 0.4 and pytorch 1.8.0.

1. On Reference platform, the initial explosion problem disappears and the simulation looks good.
2. On CUDA platform, the initial explosion problem also disappears, but I notice a disagreement between total energy and the sum of individual energy terms. The following script can show the energy disagreement problem. The problem does not appear deterministically, you might need to run it a few times to see that the total energy is not equal to the sum of the two individual energy terms.

```Python
import torch
import numpy as np
import openmm.app  as app
import openmm as omm
import openmm.unit as unit
from openmmtorch import TorchForce
import copy
from sys import exit

class NNForceDihedral(torch.nn.Module):
    def __init__(self, num_dihedrals):
        super(NNForceDihedral, self).__init__()

        self.num_dihedrals = num_dihedrals
        self.fc1 = torch.nn.Linear(2*num_dihedrals, num_dihedrals)
        self.fc2 = torch.nn.Linear(num_dihedrals, 1, bias = False)

    def forward(self, dihedrals):
        x = torch.cat((torch.cos(dihedrals), torch.sin(dihedrals)), -1)
        x = torch.tanh(self.fc1(x))
        output = self.fc2(x)
        output = torch.squeeze(output)
        return output

class ForceModule(torch.nn.Module):
    def __init__(self, num_particles):
        super(ForceModule, self).__init__()
        self.num_particles = num_particles
        self.nnforce_dihedral = NNForceDihedral(num_particles - 3)

        self.index = torch.nn.Parameter(
            torch.LongTensor( [[i, i+1, i+2, i+3] for i in range(num_particles - 3)]),
            requires_grad = False
        )

    def forward(self, positions):
        xyz = torch.unsqueeze(positions, 0)

        xyz_i = torch.index_select(xyz, 1, self.index[:, 0])
        xyz_j = torch.index_select(xyz, 1, self.index[:, 1])
        xyz_k = torch.index_select(xyz, 1, self.index[:, 2])
        xyz_l = torch.index_select(xyz, 1, self.index[:, 3])

        b1 = xyz_i - xyz_j
        b2 = xyz_j - xyz_k
        b3 = xyz_l - xyz_k

        b1_cross_b2 = torch.cross(b1, b2)
        b3_cross_b2 = torch.cross(b3, b2)

        cos_d = torch.norm(b2, dim = -1)*torch.sum(b1_cross_b2*b3_cross_b2, -1)
        sin_d = torch.sum(b2*torch.cross(b3_cross_b2, b1_cross_b2), -1)

        dihedrals = torch.atan2(sin_d, cos_d)
#        potential_energy = self.nnforce_dihedral(dihedrals)*torch.tensor(0.0)
        potential_energy = self.nnforce_dihedral(dihedrals)
        return potential_energy

xyz = torch.tensor([[ 1.357623  , -1.3806132 , -1.2257496 ],
                    [ 1.0385588 , -1.3198054 , -1.4279023 ],
                    [ 0.7484111 , -1.471794  , -1.6183405 ],
                    [ 0.6739166 , -1.2910188 , -1.9467432 ],
                    [ 0.33078226, -1.451046  , -1.9793787 ],
                    [ 0.15436828, -1.2586378 , -1.68369   ],
                    [ 0.42108235, -1.0122279 , -1.616695  ],
                    [ 0.4789558 , -1.1335697 , -1.2618917 ],
                    [ 0.8091399 , -1.080018  , -1.087149  ],
                    [ 0.97696364, -1.3592714 , -0.9014565 ]], requires_grad = True)

system_reference = omm.System()
system_cuda = omm.System()
for _ in range(10):
    system_reference.addParticle(100)
    system_cuda.addParticle(100)

harmonic_bond_force = omm.HarmonicBondForce()
for i in range(9):
    harmonic_bond_force.addBond(i, i+1, 0.38, 7e4)

harmonic_bond_force.setForceGroup(0)
system_reference.addForce(copy.copy(harmonic_bond_force))
system_cuda.addForce(copy.copy(harmonic_bond_force))

force_module = ForceModule(10)
force_module_float = ForceModule(10).float()
force_module_double = ForceModule(10).double()

scripted_force_module_float = torch.jit.script(force_module_float)
scripted_force_module_double = torch.jit.script(force_module_double)

scripted_force_module_float.save(f"./output/test/force_module_float.pth")
scripted_force_module_double.save(f"./output/test/force_module_double.pth")

nnforce_float = TorchForce(f"./output/test/force_module_float.pth")
nnforce_float.setForceGroup(1)

nnforce_double = TorchForce(f"./output/test/force_module_double.pth")
nnforce_double.setForceGroup(1)

system_reference.addForce(nnforce_double)
system_cuda.addForce(nnforce_float)

T = 300*unit.kelvin
fricCoef = 1./unit.picoseconds
stepsize = 2. * unit.femtoseconds
## simulation with reference platform
print("="*80)
print("Simulate system with reference platform")
print("="*80)
integrator = omm.LangevinMiddleIntegrator(T, fricCoef, stepsize)
platform = omm.Platform.getPlatformByName('Reference')
context = omm.Context(system_reference, integrator, platform)
context.setPositions(xyz.data.numpy())

for i in range(1000):
    integrator.step(1)

    state = context.getState(getEnergy = True, getForces = True, groups = set([0]))
    bond_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    bond_force = state.getForces()
    bond_force = np.array(bond_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    state = context.getState(getEnergy = True, getForces = True, groups = set([1]))
    torch_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    torch_force = state.getForces()
    torch_force = np.array(torch_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    state = context.getState(getEnergy = True, getForces = True)
    total_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    total_force = state.getForces()
    total_force = np.array(total_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))
    assert(abs(total_energy - bond_energy - torch_energy) < 1e-5)

    if (i + 1) % 50 == 0:
        print(f"step: {i:>5}, bond_energy: {bond_energy:>10.2f} kJ/mol, torch_energy: {torch_energy:>10.2f} kJ/mol, bond_force[0,0]:", bond_force[0,0], 'torch_force[0,0]: ', torch_force[0,0], f'total_energy: {total_energy:>10.2f} kJ/mol', 'total_force[0,0]: ', total_force[0,0])

## simulation with CUDA platform
print("="*80)
print("Simulate system with CUDA platform")
print("="*80)
integrator = omm.LangevinMiddleIntegrator(T, fricCoef, stepsize)
platform = omm.Platform.getPlatformByName('CUDA')
context = omm.Context(system_cuda, integrator, platform)
context.setPositions(xyz.data.numpy())

for i in range(1000):
    integrator.step(1)

    state = context.getState(getEnergy = True, getForces = True, groups = set([0]))
    bond_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    bond_force = state.getForces()
    bond_force = np.array(bond_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    state = context.getState(getEnergy = True, getForces = True, groups = set([1]))
    torch_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    torch_force = state.getForces()
    torch_force = np.array(torch_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    state = context.getState(getEnergy = True, getForces = True)
    total_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    total_force = state.getForces()
    total_force = np.array(total_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    #assert(abs(total_energy - bond_energy - torch_energy) < 1e-3)

    if abs(total_energy - bond_energy - torch_energy) > 1e-3:
        print(f"step: {i:>5}, bond_energy: {bond_energy:>10.4f} kJ/mol, torch_energy: {torch_energy:>10.4f} kJ/mol, total_energy: {total_energy:>10.4f} kJ/mol")
        exit()

    if (i + 1) % 50 == 0:
        print(f"step: {i:>5}, bond_energy: {bond_energy:>10.2f} kJ/mol, torch_energy: {torch_energy:>10.2f} kJ/mol, bond_force[0,0]:", bond_force[0,0], 'torch_force[0,0]: ', torch_force[0,0], f'total_energy: {total_energy:>10.2f} kJ/mol', 'total_force[0,0]: ', total_force[0,0])

exit()
```

**Comment by raimis** [comment]
@xqding thanks for the second test.

Good news: https://github.com/openmm/openmm/pull/3256 solves the problem, but it isn't not the best solution.
Bad news: I tried the better solution (https://github.com/openmm/openmm/pull/3258), but it doesn't work. It needs some modifications on OpenMM-Torch side.

**Comment by raimis** [comment]
This is a simplified script which triggers the same issue:
```python
import sys
import torch
import numpy as np
import simtk.openmm as omm
import simtk.unit as unit
from openmmtorch import TorchForce

class Model(torch.nn.Module):
    def __init__(self, num_particles):
        super().__init__()
        self.index = torch.nn.Parameter(torch.LongTensor([[i, i+1, i+2, i+3] for i in range(num_particles - 3)]), requires_grad=False)
        self.fc = torch.nn.Linear(2*len(self.index), 1)

    def forward(self, positions):
        xyz = torch.unsqueeze(positions, 0)

        xyz_i = torch.index_select(xyz, 1, self.index[:, 0])
        xyz_j = torch.index_select(xyz, 1, self.index[:, 1])
        xyz_k = torch.index_select(xyz, 1, self.index[:, 2])
        xyz_l = torch.index_select(xyz, 1, self.index[:, 3])

        b1 = xyz_i - xyz_j
        b2 = xyz_j - xyz_k
        b3 = xyz_l - xyz_k

        b1_cross_b2 = torch.cross(b1, b2)
        b3_cross_b2 = torch.cross(b3, b2)

        cos_d = torch.norm(b2, -1) * torch.sum(b1_cross_b2*b3_cross_b2, -1)
        sin_d = torch.sum(b2 * torch.cross(b3_cross_b2, b1_cross_b2), -1)

        dihedrals = torch.atan2(sin_d, cos_d)
        cos_sines = torch.cat((torch.cos(dihedrals), torch.sin(dihedrals)), -1)

        return self.fc(cos_sines)

num_particles = 4

system = omm.System()
for _ in range(num_particles):
    system.addParticle(100)

force = omm.HarmonicBondForce()
for i in range(num_particles-1):
    force.addBond(i, i+1, 1, 1)
force.setForceGroup(0)
system.addForce(force)

torch.jit.script(Model(num_particles)).save("model.pth")
force = TorchForce("model.pth")
force.setForceGroup(1)
system.addForce(force)

integrator = omm.VerletIntegrator(1*unit.femtoseconds)
platform = omm.Platform.getPlatformByName('CUDA')
context = omm.Context(system, integrator, platform)
context.setPositions(np.random.rand(num_particles, 3))

for i in range(10000):
    integrator.step(1)

    state = context.getState(getEnergy = True, getForces = True, groups = set([0]))
    bond_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    bond_force = state.getForces()
    bond_force = np.array(bond_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    state = context.getState(getEnergy = True, getForces = True, groups = set([1]))
    torch_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    torch_force = state.getForces()
    torch_force = np.array(torch_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    state = context.getState(getEnergy = True, getForces = True)
    total_energy = state.getPotentialEnergy().value_in_unit(unit.kilojoule_per_mole)
    total_force = state.getForces()
    total_force = np.array(total_force.value_in_unit(unit.kilojoule_per_mole/unit.nanometer))

    if (i + 1) % 100 == 0:
        print(f"step: {i:>5}, bond_energy: {bond_energy:>10.4f} kJ/mol, torch_energy: {torch_energy:>10.4f} kJ/mol, total_energy: {total_energy:>10.4f} kJ/mol")

    if abs(total_energy - bond_energy - torch_energy) > 1e-3:
        print(f"step: {i:>5}, bond_energy: {bond_energy:>10.4f} kJ/mol, torch_energy: {torch_energy:>10.4f} kJ/mol, total_energy: {total_energy:>10.4f} kJ/mol")
        sys.exit()
```

**Comment by peastman** [comment]
The above script reports no errors when I run it against the current main branches for OpenMM and OpenMM-Torch, that is, with no fixes for CUDA context management.  I ran it three times to be sure.  It successfully got through the 10,000 steps all three times.  That's with PyTorch 1.7.1.

**Comment by raimis** [comment]
@peastman does the original @xqding script (https://github.com/openmm/openmm-torch/issues/42#issuecomment-931386518) reports an error?

**Comment by peastman** [comment]
I can reproduce it with the original script.  And I verified my fix works.  I'll post a PR.

**Comment by peastman** [comment]
The fix is in #47.

**Comment by raimis** [comment]
OK! I'll test in my case.
